Base/QTGUI/qSlicerSingletonViewFactory.h

```diff
@@ -0,0 +1,68 @@
+/*==============================================================================
+
+Copyright (c) Laboratory for Percutaneous Surgery (PerkLab)
+Queen's University, Kingston, ON, Canada. All Rights Reserved.
+
+See COPYRIGHT.txt
+or http://www.slicer.org/copyright/copyright.txt for details.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+This file was originally developed by Kyle Sunderland, PerkLab, Queen's University
+and was supported through CANARIE's Research Software Program, and Cancer
+Care Ontario.
+
+==============================================================================*/
+
+#ifndef qSlicerSingletonViewFactory_h
+#define qSlicerSingletonViewFactory_h
+
+// CTK includes
+#include "ctkLayoutViewFactory.h"
+
+// QtGUI includes
+#include "qSlicerBaseQTGUIExport.h"
+
+class ctkDICOMBrowser;
+class qSlicerSingletonViewFactoryPrivate;
+
+/// This class provides an interface for C++ and Python classes to register a singleton view replacement widget.
+/// New view widgets can be registered by registering a new qSlicerSingletonViewFactory, and then setting the widget and tag using setWidget(QWidget*) and
+/// setTagName(QString).
+/// This factory contains a single pointer to an instance of the widget, so only one view can be created within a given layout.
+class Q_SLICER_BASE_QTGUI_EXPORT qSlicerSingletonViewFactory : public ctkLayoutViewFactory
+{
+  Q_OBJECT
+public:
+  typedef ctkLayoutViewFactory Superclass;
+  qSlicerSingletonViewFactory(QObject* parent=0);
+  virtual ~qSlicerSingletonViewFactory();
+
+  /// Reimplemented to support custom element names
+  virtual QStringList supportedElementNames()const;
+
+  /// Set the singleton widget instance that will be used to create the view
+  Q_INVOKABLE virtual void setWidget(QWidget* widget);
```

Comment:
Who is responsible for deleting the widget? It should be at least clarified in the documentation. We may consider allowing the factory taking responsibility for deleting the widget if that simplifies the implementation.

Reply:
I changed the regular pointer to a QSharedPointer.
Factory is now responsible for deleting the widget.